sdk/rust/src/contract/contract_function_parameters.rs

```diff
@@ -0,0 +1,1268 @@
+        (false, self.to_be_bytes().to_vec())
+    }
+}
+
+impl IntEncode for i64 {
+    fn get_is_negative_and_be_bytes(&self) -> (bool, Vec<u8>) {
+        (self.is_negative(), self.to_be_bytes().to_vec())
+    }
+}
+
+impl IntEncode for u128 {
+    fn get_is_negative_and_be_bytes(&self) -> (bool, Vec<u8>) {
+        (false, self.to_be_bytes().to_vec())
+    }
+}
+
+impl IntEncode for i128 {
+    fn get_is_negative_and_be_bytes(&self) -> (bool, Vec<u8>) {
+        (self.is_negative(), self.to_be_bytes().to_vec())
+    }
+}
+
+impl IntEncode for BigUint {
+    fn get_is_negative_and_be_bytes(&self) -> (bool, Vec<u8>) {
+        (false, self.to_bytes_be())
+    }
+}
+
+impl IntEncode for BigInt {
+    fn get_is_negative_and_be_bytes(&self) -> (bool, Vec<u8>) {
+        (self.sign() == Sign::Minus, self.to_signed_bytes_be())
+    }
+}
+
+impl ContractFunctionParameters {
+    /// Create a new, empty `ContractFunctionParameters`
+    #[allow(dead_code)]
+    pub fn new() -> Self {
+        ContractFunctionParameters { args: vec![] }
+    }
+
+    fn left_pad_32_bytes(bytes: &[u8], is_negative: bool) -> Vec<u8> {
+        let pad_byte = if is_negative { 0xFF } else { 0x00 };
+        iter::repeat(pad_byte).take(32 - bytes.len()).chain(bytes.iter().map(|b| *b)).collect()
+    }
+
+    fn right_pad_32_bytes(bytes: &[u8]) -> Vec<u8> {
+        bytes.iter().map(|b| *b).chain(iter::repeat(0x00).take(32 - (bytes.len() % 32))).collect()
+    }
+
+    fn encode_address(address_string: &str) -> Vec<u8> {
+        Self::left_pad_32_bytes(
+            IdEvmAddress::from_str(address_string).unwrap().0 .0.as_slice(),
+            false,
+        )
+    }
+
+    fn encode_dynamic_bytes(bytes: &[u8]) -> Vec<u8> {
+        let mut out_bytes = Self::left_pad_32_bytes(bytes.len().to_be_bytes().as_slice(), false);
+        out_bytes.append(&mut Self::right_pad_32_bytes(bytes));
+        out_bytes
+    }
+
+    fn encode_array_of_dynamic_byte_arrays<I>(elements: I, elements_len: usize) -> Vec<u8>
+    where
+        I: IntoIterator,
+        I::Item: AsRef<[u8]>,
+    {
+        let head_len = (elements_len + 1) * 32;
+        let mut out_bytes = Vec::with_capacity(head_len);
+        out_bytes
+            .append(&mut Self::left_pad_32_bytes(elements_len.to_be_bytes().as_slice(), false));
+        let mut current_offset = elements_len * 32;
+        let mut body_bytes: Vec<u8> = Vec::new();
+        for element in elements {
+            let element = element.as_ref();
+            out_bytes.append(&mut Self::left_pad_32_bytes(
+                current_offset.to_be_bytes().as_slice(),
+                false,
+            ));
+            current_offset += element.len();
+            body_bytes.extend(element)
+        }
+        out_bytes.extend(body_bytes);
+        out_bytes
+    }
+
+    fn encode_array_of_32_byte_elements<I>(elements: I, elements_len: usize) -> Vec<u8>
+    where
+        I: IntoIterator<Item = Vec<u8>>,
```

Comment:
This feels like it should be 
```suggestion
    fn encode_array_of_32_byte_elements<T: AsRef<[u8]>>(elements: &[T]) -> Vec<u8>
```
Because, if you need to get the exact length you should either be using `TrustedLen` (an unsafe iterator trait that's also only available in the standard library itself, it basically says "unsafe code can trust this length"), `ExactSizeIterator` (which exists and you can use, just not unsafely assume, it would work here), or a slice.

So, the other option is
```suggestion
    fn encode_array_of_32_byte_elements<I, U>(elements: I) -> Vec<u8>
    where
        I: IntoIterator<Item = U>,
        I::IntoIter: ExactSizeIterator,
        U: AsRef<[u8]>
```
with either of these, the elements would be `elements.into_iter().flat_map(|it| it.as_ref().iter().copied())`
note: if you go with slices, use `iter` instead of `into_iter`
note2: If you had `left_pad_32_bytes` and `truncate_and_left_pad_32_bytes` return an array instead use `[u8; 32]` instead of `AsRef<[u8]>`, `[u8; 32]` does implement `Copy` so if you get errors about needing `[u8; 32]` when you have `&[u8; 32]` you'd be able to do `.copied()` on the iterators.

Reply:
I tried messing around with these, and my verdict is... nope.